File: kpis/kpiBrakeMode.m
function kpiBrakeMode(obj, i, aebStartIdx)
    % Determine if PB and FB are activated during AEB event
    % Assuming PB is indicated by aebTargetDecel dropping below -6 and FB below -12
    %% Inputs:
    % obj            - KPIExtractor object
    % i              - Index of the file in kpiTable
    % aebStartIdx    - Index of AEB request event

    % Use obj.kpiTable and obj.signalMatChunk directly
    kpiTable        = obj.kpiTable;
    signalMatChunk  = obj.signalMatChunk;
    PB_TGT_DECEL    = obj.PB_TGT_DECEL;
    FB_TGT_DECEL    = obj.FB_TGT_DECEL;
    TGT_TOL         = obj.TGT_TOL;

    aebEndReq = length(signalMatChunk.time); % Simplified: you can reuse your end-detection logic

    % Ensure required columns exist
    if ~ismember('pbDur', kpiTable.Properties.VariableNames)
        obj.kpiTable.pbDur = NaN(height(obj.kpiTable), 1);
    end
    if ~ismember('fbDur', kpiTable.Properties.VariableNames)
        obj.kpiTable.fbDur = NaN(height(obj.kpiTable), 1);
    end
    if ~ismember('isPBOn', kpiTable.Properties.VariableNames)
        obj.kpiTable.isPBOn = false(height(obj.kpiTable), 1);
    end
    if ~ismember('isFBOn', kpiTable.Properties.VariableNames)
        obj.kpiTable.isFBOn = false(height(obj.kpiTable), 1);
    end

    % Always check from AEB request to the end of recording
    segment = signalMatChunk.aebTargetDecel(aebStartIdx:aebEndReq);

    % Find all PB / FB candidate indices
    pbIdx = find(abs(segment - PB_TGT_DECEL) <= TGT_TOL);
    fbIdx = find(abs(segment - FB_TGT_DECEL) <= TGT_TOL);

    % PB and FB status
    isPBOn = ~isempty(pbIdx);
    isFBOn = ~isempty(fbIdx);
    
    % Durations in seconds
    if isPBOn
        pbStart = aebStartIdx + pbIdx(1) - 1;   % Convert back to global index
        pbEnd   = aebStartIdx + pbIdx(end) - 1;
        pbDur = signalMatChunk.time(pbEnd) - signalMatChunk.time(pbStart);
        if isduration(pbDur)
            pbDur_value = seconds(pbDur);
        else
            pbDur_value = double(pbDur); % Handle numeric input
        end
        if ~isnan(pbDur_value) && isfinite(pbDur_value)
            obj.kpiTable.pbDur(i) = round(pbDur_value, 2);
        else
            warning('Invalid pbDur value for file index %d: %s. Setting pbDur to 0.', i, string(pbDur));
            obj.kpiTable.pbDur(i) = 0;
        end
    else
        obj.kpiTable.pbDur(i) = 0;
    end

    if isFBOn
        fbStart = aebStartIdx + fbIdx(1) - 1;
        fbEnd   = aebStartIdx + fbIdx(end) - 1;
        fbDur = signalMatChunk.time(fbEnd) - signalMatChunk.time(fbStart);
        if isduration(fbDur)
            fbDur_value = seconds(fbDur);
        else
            fbDur_value = double(fbDur); % Handle numeric input
        end
        if ~isnan(fbDur_value) && isfinite(fbDur_value)
            obj.kpiTable.fbDur(i) = round(fbDur_value, 2);
        else
            warning('Invalid fbDur value for file index %d: %s. Setting fbDur to 0.', i, string(fbDur));
            obj.kpiTable.fbDur(i) = 0;
        end
    else
        obj.kpiTable.fbDur(i) = 0;
    end

    obj.kpiTable.isPBOn(i) = isPBOn;
    obj.kpiTable.isFBOn(i) = isFBOn;
end

File: python/kpis/aeb/kpi_brake_mode.py
import numpy as np
import pandas as pd
import warnings

def kpi_brake_mode(obj, i, aeb_start_idx):
    """
    Determine if PB and FB are activated during AEB event.
    Updates obj.kpi_table in-place.

    Parameters
    ----------
    obj : object with attributes
        - kpi_table : pandas.DataFrame
        - signal_mat_chunk : dict-like with arrays {time, aebTargetDecel}
        - PB_TGT_DECEL : float
        - FB_TGT_DECEL : float
        - TGT_TOL : float
    i : int
        Index of the file in kpi_table.
    aeb_start_idx : int
        Index of AEB request event.
    """
    kpi_table = obj.kpi_table
    signals = obj.signal_mat_chunk
    PB_TGT_DECEL = obj.PB_TGT_DECEL
    FB_TGT_DECEL = obj.FB_TGT_DECEL
    TGT_TOL = obj.TGT_TOL

    aeb_end_req = len(signals["time"]) - 1  # equivalent to MATLAB length()

    # Ensure required columns exist
    for col, default, dtype in [
        ("pbDur", np.nan, "float"),
        ("fbDur", np.nan, "float"),
        ("isPBOn", False, "bool"),
        ("isFBOn", False, "bool"),
    ]:
        if col not in kpi_table.columns:
            kpi_table[col] = pd.Series([default] * len(kpi_table), dtype=dtype)

    # Segment from AEB start to end
    segment = signals["aebTargetDecel"][aeb_start_idx:aeb_end_req + 1]

    # Find PB / FB candidate indices
    pb_idx = np.where(np.abs(segment - PB_TGT_DECEL) <= TGT_TOL)[0]
    fb_idx = np.where(np.abs(segment - FB_TGT_DECEL) <= TGT_TOL)[0]

    is_pb_on = len(pb_idx) > 0
    is_fb_on = len(fb_idx) > 0

    # PB duration
    if is_pb_on:
        pb_start = aeb_start_idx + pb_idx[0]
        pb_end = aeb_start_idx + pb_idx[-1]
        pb_dur = signals["time"][pb_end] - signals["time"][pb_start]

        pb_dur_value = float(pb_dur)
        if np.isfinite(pb_dur_value):
            kpi_table.at[i, "pbDur"] = round(pb_dur_value, 2)
        else:
            warnings.warn(f"Invalid pbDur value for file index {i}: {pb_dur}. Setting pbDur to 0.")
            kpi_table.at[i, "pbDur"] = 0.0
    else:
        kpi_table.at[i, "pbDur"] = 0.0

    # FB duration
    if is_fb_on:
        fb_start = aeb_start_idx + fb_idx[0]
        fb_end = aeb_start_idx + fb_idx[-1]
        fb_dur = signals["time"][fb_end] - signals["time"][fb_start]

        fb_dur_value = float(fb_dur)
        if np.isfinite(fb_dur_value):
            kpi_table.at[i, "fbDur"] = round(fb_dur_value, 2)
        else:
            warnings.warn(f"Invalid fbDur value for file index {i}: {fb_dur}. Setting fbDur to 0.")
            kpi_table.at[i, "fbDur"] = 0.0
    else:
        kpi_table.at[i, "fbDur"] = 0.0

    # Flags
    kpi_table.at[i, "isPBOn"] = is_pb_on
    kpi_table.at[i, "isFBOn"] = is_fb_on
